### ISSUE TITLE
Large sprite and multihex mountain layering glitch

### ISSUE THREAD

**Issue by doofus-01** [wesnoth/wesnoth#6203]
### Game and System Information
 - What version of the game are you running?
 Battle for Wesnoth v1.15.18+dev (6f499ddeabd-Modified) x86_64

### Describe the bug
Large sprites bleed into central section of mountains.  This was reported on the forums: https://forums.wesnoth.org/viewtopic.php?p=668409#p668409

### To Reproduce

1. In scenario editor, make some mountains that use the basic5_*.png or the basic_range3_*.png
2. Place some direwolves above and to the leftof those mountain peaks.

### Expected behavior
The direwolf is behind the mountain peak, the paws don't peak out along the edge of the hex.

**Comment by doofus-01** [comment]
I looked into the original issue more, and conclude that this can't be fixed with terrain-graphics configurations.  We know the firedragon sprite breaks a lot of things, the direwolf is just a more subtle version of the same thing.  Maybe keeping big sprites from crossing the SW-S-SE hex borders will help, but I haven't tested that yet.

The image here is wolves bleeding through a mountain at the hex edges, but the mountain image (with some yellow bars to make sure) is one big one, not the constructed version in the core MOUNTAINS_2x2 instance.  Mountain on the left has a different-by-1 `base` key value than the one on the right, showing that simple change won't help.
![Screenshot_20211218](https://user-images.githubusercontent.com/13510196/146659259-87809ed6-603a-484c-b062-2a72be2df08e.jpg)

The example posted by max-torch is doesn't look like the same thing, it might be fixable.

**Comment by CelticMinstrel** [comment]
It seems to me that, nearly always, you'd want all sprites to be unconditionally above all terrain… so the logical thing to do would be to draw all the terrain before drawing any of the sprites, which I'm pretty sure would solve this issue. (It's probably not enough by itself, as there are no doubt some unusual cases where you _do_ want terrain above units… perhaps the sails on the boat in SotA? I'm not sure though.)

If we _must_ have terrain and unit sprites fit into the same layering system, then at the very least the procedure should be "draw everything on layer 1", "draw everything on layer 2", etc, rather than "draw everything on hex 1", "draw everything on hex 2", etc. I think that would also solve this issue without breaking unusual cases.

**Comment by doofus-01** [comment]
The layer system already exists.  The base system is needed for the 2.5D effect.  Wesnoth could mimic a board game, where sprites and map don't interact at all, but that would be a significant change.  I would be 100% opposed to that change.

**Comment by CelticMinstrel** [comment]
I know the layer system exists, but how exactly does it work? It seems to me that the correct answer would be to draw everything in one layer on every hex, then everything in the next layer on every hex, and so forth, whereas the incorrect answer would be to draw everything in every layer on one hex, then everything in every layer on the next hex, and so forth… and the latter method seems like a possible cause for this issue. This is just speculation though, as I don't know how the code actually works.

Although… I think you could also make an argument that the image in the original post is actually intended behaviour; that mountain really _is_ meant to be in front of the wolf. Your most recent screenshot does clearly have errors though, even assuming that's the intent.

**Comment by doofus-01** [comment]
Right, the base key allows you to have the mountain above or below the sprite depending on their vertical (y) positions.

**Comment by cooljeanius** [comment]
I wonder if this terrain graphics glitch I noticed with some custom terrain in Flight to Freedom is also an example of this bug? It also has to do with multihex mountain layering: 
<img width="181" alt="Screen Shot 2022-11-27 at 20 04 04" src="https://user-images.githubusercontent.com/1388660/215241423-05c614f5-51cc-48fd-8600-cd84974bbd8b.png">

**Comment by doofus-01** [comment]
No idea, could be at least a couple things.  Depending on what you're doing with that little chasm, you might be able to just raise the layer it's drawn on, or lower the `base` key's coordinates.

**Comment by doofus-01** [comment]
I can reproduce this in 1.16, but enough has changed in 1.17 that this is no longer an issue going forward.

**Comment by Wedge009** [comment]
Can close when 1.18 releases, then?

**Comment by doofus-01** [comment]
I'm not sure what my April 2023 comment was about (might have been because of a sprite-above-all-terrain bug that has since been fixed), but I can reproduce this in current master.

**Comment by mentos987** [comment]
Probably fixed in https://github.com/wesnoth/wesnoth/pull/10702, not entirely sure.

No, it is not a fix. To fix this we should use y offset to move the images up by 5 pixels. The real issue is that normal sized units are not ever supposed to intrude on any of the three hexes underneath it, which the wolf does.

It is fine to intrude on the three hexes above though.

In order to fully implement offsets we need to add the feature requested by: https://github.com/wesnoth/wesnoth/issues/9746 first.

@soliton- things are coming full circle.

**Comment by mentos987** [comment]
Here is the fix: https://github.com/wesnoth/wesnoth/pull/10707

We will need to go through all offenders and adjust their unit.cfg files.

**Comment by cooljeanius** [comment]
> I wonder if this terrain graphics glitch I noticed with some custom terrain in Flight to Freedom is also an example of this bug? It also has to do with multihex mountain layering: 
> <img alt="Screen Shot 2022-11-27 at 20 04 04" width="181" src="https://user-images.githubusercontent.com/1388660/215241423-05c614f5-51cc-48fd-8600-cd84974bbd8b.png">

@MadMax3947

**Comment by mentos987** [comment]
The custom mountain terrain you are using is placed on terrain background level. The mountain tops that is "glitching" for you are part of terrain_foreground layer. 

I am currently writing another PR that lets you change the layer of an image. But keep in mind that if you change something to terrain foreground it will appear above units. So your gorge would look terrible if you moved units on those tiles. 

@cooljeanius 

What you have isn't really a glitch as is. And there is no easy way to fix it. 

Another thing I am working on is multi-hex images. So you could to a print screen and adjust the image and then place that print screen on top of the entire area as a multi-hex image. You would need to delete the underlying mountains to prevent the terrain_foreground stuff from poking through though.

If you set the gorge area as impassable you can simply put it all as a halo too. As long as no unit is near it it will look fine.
